program-analysis/echidna/advanced/state-network-forking.md

```diff
@@ -0,0 +1,96 @@
+# On-chain fuzzing with state forking
+
+**Table of contents:**
+
+- [On-chain fuzzing with state forking](#on-chain-fuzzing-with-state-forking)
+  - [Example](#example)
+  - [Corpus and RPC cache](#corpus-and-rpc-cache)
+  - [Coverage and Etherscan integration](#coverage-and-etherscan-integration)
+
+## Example
+
+One of the most anticipated features of Echidna 2.1.0 is the state network forking. This means that Echidna can run starting with an existing blockchain state provided by an external RPC service (Infura, Alchemy, local node, etc). 
+This enables users to speed up the fuzzing setup when using already deployed contracts. For instance:
+
+```solidity
+interface IHevm {
+    function warp(uint256 newTimestamp) external;
+    function roll(uint256 newNumber) external;
+}
+
+interface Compound {
+  function mint() external payable;
+  function balanceOf(address) external view returns (uint256);
+}
+
+contract TestCompoundEthMint {
+  address constant HEVM_ADDRESS = 0x7109709ECfa91a80626fF3989D68f67F5b1DD12D;
+  IHevm hevm = IHevm(HEVM_ADDRESS);
+  Compound comp = Compound(0x4Ddc2D193948926D02f9B1fE9e1daa0718270ED5);
+
+  constructor() {
+    hevm.roll(16771449);  // sets the correct block number
```

Comment:
Does this mean that by writing `hevm.roll(16771449);`, you will fork the blockchain state from block 16771449? This behavior would then be different to the way foundry works. There an initial block number parameter is passed in and if you skip to a new block, it's all just simulated locally, but your state is still loaded from the initially set block height.

This difference is not necessarily bad, but it could break some tests that rely on relative time/block jumps, i.e. the contract requires `10` blocks to have passed and you write `hevm.roll(block.number + 10)`, but now the actual state has changed and your assumptions on the current state might be different.

Reply:
I added a note about this.